File: src/robot_model_loader.cpp
#include "torch-moveit.h"
#include "utils.h"
#include <moveit/robot_model_loader/robot_model_loader.h>

typedef robot_model_loader::RobotModelLoaderPtr RobotModelLoaderPtr;
typedef robot_model::RobotModelPtr RobotModelPtr;

MOVIMP(RobotModelLoaderPtr*, RobotModelLoader, new)(
  const char *robot_description /*= "robot_description"*/,
  bool load_kinematics_solvers /*= true*/
)
{
   return new RobotModelLoaderPtr(new robot_model_loader::RobotModelLoader(std::string(robot_description), load_kinematics_solvers));
}

MOVIMP(void, RobotModelLoader, delete)(RobotModelLoaderPtr *ptr)
{
  if (ptr)
    delete ptr;
}

MOVIMP(void, RobotModelLoader, release)(RobotModelLoaderPtr *ptr)
{
  ptr->reset();
}

MOVIMP(void, RobotModelLoader, getModel)(RobotModelLoaderPtr *ptr, RobotModelPtr *output)
{
  *output = (*ptr)->getModel();
}

MOVIMP(const char *, RobotModelLoader, getRobotDescription)(RobotModelLoaderPtr *ptr)
{
  return (*ptr)->getRobotDescription().c_str();
}

File: RobotModelLoader.lua
local ffi = require 'ffi'
local torch = require 'torch'
local moveit = require 'moveit.env'
local utils = require 'moveit.utils'
local ros = require 'ros'
local std = ros.std

local RobotModelLoader = torch.class('moveit.RobotModelLoader', moveit)

local f

function init()
  local RobotModelLoader_method_names = {
    "new",
    "delete",
    "release",
    "getModel",
    "getRobotDescription"
  }
  f = utils.create_method_table("moveit_RobotModelLoader_", RobotModelLoader_method_names)
end

init()

function RobotModelLoader:__init(robot_description, load_kinematics_solvers)
  self.o = f.new(robot_description or "robot_description", load_kinematics_solvers or true)
end

function RobotModelLoader:cdata()
  return self.o
end

function RobotModelLoader:release()
  f.release(self.o)
end

function RobotModelLoader:getModel()
  local model = moveit.RobotModel()
  f.getModel(self.o, model:cdata())
  return model
end

function RobotModelLoader:getRobotDescription()
  return ffi.string(f.getRobotDescription(self.o))
end
